google-cloud-spanner/src/main/java/com/google/cloud/spanner/ThreadFactoryUtil.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import com.google.api.core.InternalApi;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import javax.annotation.Nullable;
+
+/** Utility class for creating a thread factory for daemon or virtual threads. */
+@InternalApi
+public class ThreadFactoryUtil {
+
+  /**
+   * Tries to create a thread factory for virtual threads, and otherwise falls back to creating a
+   * platform thread factory that creates daemon threads. Virtual threads are supported from JDK21.
+   *
+   * @param baseNameFormat the base name format for the threads, '-%d' will be appended to the
+   *     actual thread name format
+   * @param tryVirtualThreads whether to try to use virtual threads if available or not
+   * @return a {@link ThreadFactory} that produces virtual threads (Java 21 or higher) or daemon
+   *     threads
+   */
+  @InternalApi
+  public static ThreadFactory createVirtualOrDaemonThreadFactory(
```

Comment:
Aren't virtual thread always daemon? Should we rename to createVirtualOrPlatformThreadFactory and also replace daemon by platform throughout PR if true

Reply:
Virtual threads are indeed always daemon threads. The reason that I chose this name is to indicate that the thread factory will create either virtual or daemon (platform) threads. Put another way: Calling the method `createVirtualOrPlatformThreadFactory` does not indicate that the platform threads will be daemon threads. I renamed it to `createVirtualOrPlatformDaemonThreadFactory`